docs/architecture/ADR-003-borrow-assets.md

```diff
@@ -0,0 +1,125 @@
+
+Additionally it has been discussed that, rather than an account's specific uToken denoms being tied to specific borrow positions, the sum of all borrow positions and collateral uTokens related to an account is used to check the account's overcollateralization levels and mark its positions for liquidation.
+
+Note that the exchange rate of Assets:uAssets (e.g. `uatom:u/uatom`) is still a shifting exchange rate that grows with interest - see [ADR-001: Interest Stream](./ADR-001-interest-stream.md).
+
+In contrast, the exchange rate of original assets:borrowed assets (e.g. `uatom:ether`) can only be determined using a price oracle.
+
+The calculated borrow limit, which weighs collateral uTokens against borrowed assets (e.g. `u/uatom:ether`) is derived from combining the two above.
+
+Note also that as a consequence of uToken interest, the asset value of uToken collateral increases over time, meaning a user who repays positions in full and redeems collateral uTokens will receive back more base assets than they deposited originally, reducing effective interest.
+
+## Detailed Design
+
+For the purposes of borrowing and repaying assets, as well as marking uTokens as collateral, the `umee/x/leverage` module does not mint or burn tokens. It uses its keeper to remember open positions and user collateral settings, and the `x/bank` module to perform all necessary balance checks and token transfers.
+
+Additionally, the `x/bank` module must be extended to prevent users from transferring uTokens when doing so would lower their borrow limit below thir current total borrowed amount.
+
+Because borrow limits weight the value of different token types together, the calculation will require the use of price oracles (and placeholder functions before those are available).
+
+### Basic Message Types
+
+To implement the borrow/repay/collateral functionality of the Asset Facility, the four common message types will be:
+```go
+// MsgSetCollateral - a borrower enables or disables a specific utoken type in their wallet to be used as collateral
+type MsgSetCollateral struct {
+  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
+  Denom    string         `json:"denom" yaml:"denom"`
+  Enable   bool           `json:"enable" yaml:"enable"`
+}
+// MsgBorrowAsset - a user wishes to borrow assets of one or more allowed types
+type MsgBorrowAsset struct {
+  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
+  Amount   sdk.Coins      `json:"amount" yaml:"amount"`
+}
+// MsgRepayAsset - a user wishes to repay assets of one or more borrowed types
+type MsgRepayAsset struct {
+  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
+  Amount   sdk.Coins      `json:"amount" yaml:"amount"`
+}
+```
+Messages must use only whitelisted denominations. Collateral is always a uToken denomination, and assets are never uTokens.
```

Comment:
I personally do not care, but I think some folks might get offended by certain nomenclature, so we should stick with "allow-list" and "deny-list" instead of whitelist/blacklist 👍 

```suggestion
Messages must use denominations only in the allow-list. Collateral is always a uToken denomination, and assets are never uTokens.
```